### PR TITLE
Feature/task batches api

### PR DESCRIPTION
**Pull request by DeFidelity**
Completed the spec requirements for task batches API, referenced here: https://github.com/turian/audio-discrimination-crowdsource/issues/31#issuecomment-1246358198

Closes #76

### CODE REVIEW

**Comment by turian** [comment]
Screenshots must be included in any PR where relevant

**Comment by DeFidelity** [comment]
> 



> Screenshots must be included in any PR where relevant

Added screenshot

**Comment by turian** [comment]
@DeFidelity Where is the screenshot of the web app with the updated functionnality?

File: polls/serializers.py [hunk 1]
```diff
@@ -12,3 +12,20 @@ class TaskSerializer(ModelSerializer):
     class Meta:
         model = Task
         exclude = ["batch"]
+
+
+class BatchTaskSerializer(ModelSerializer):
```
Comment (maulikkanani2) [inline polls/serializers.py]: BatchTaskSerializer is used to create a Batch object.

**Review by maulikkanani2 — APPROVED** [review]

**Review by maulikkanani2 — APPROVED** [review]

**Review by maulikkanani2 — APPROVED** [review]

**Review by maulikkanani2 — APPROVED** [review]

**Review by maulikkanani2 — APPROVED** [review]

File: polls/templates/polls/auth_token.html [hunk 2]
```diff
@@ -75,6 +78,59 @@ <h4>Lock Users API</h4>
...
+            "reference_url": "https://www.test-api1.com",
+            "transform_url": "https://www.test-api1.com",
+            "transform_metadata": {}
+        }
+    ]
+}
+
+response = requests.post(batch_task_api, headers=headers, json=payload)
+print(response.json())
 print(response.status_code)
```
Comment (VivekWangoes) [inline polls/templates/polls/auth_token.html:134]: There is no need to print.

File: polls/serializers.py [hunk 3]
```diff
@@ -12,3 +12,20 @@ class TaskSerializer(ModelSerializer):
...
+        fields = "__all__"
+
+    def create(self, validated_data):
+        tasks = validated_data.pop("tasks", [])
+        batch = Batch.objects.create(**validated_data)
+
+        for task_dict in tasks:
+            task_dict["batch"] = batch
+            Task.objects.create(**task_dict)
+        return batch
```
Comment (VivekWangoes) [inline polls/serializers.py:45]: There is no comment.

File: polls/tests.py [hunk 4]
```diff
@@ -1,6 +1,10 @@
+<<<<<<< HEAD
 from django.contrib.auth import get_user_model
 from django.urls import reverse
 from django.utils import timezone
+=======
+# from django.test import TestCase
```
Comment (VivekWangoes) [inline polls/tests.py]: Please remove the conflict code.

**Comment by DeFidelity** [comment]
@najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the respond code keeps returning 200 instead of 201.

**Comment by najamF18** [comment]
> @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.

I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201

File: polls/templates/polls/auth_token.html [hunk 5]
```diff
@@ -33,7 +33,7 @@ <h4 class="mt-3">Admin API</h4>
 headers = {
     "Authorization":admin_api_url "Token {{ token.key }}"
     }
-url = "{{ admin_api }}"
+url = "{{ admin_api_url }}"
 response = requests.get(url, headers=headers)
 print(response.json())
```
Comment (VivekWangoes) [inline polls/templates/polls/auth_token.html:38]: Please  remove the print if no need.

**Comment by najamF18** [comment]
> > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> 
> I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201

I have figured out the issue. Actually we are using a rest API for this test. And we should not be using django contrib's LoginRequiredMixin and UserPassesTestMixin. It created a conflict while processing the request and that is why the test is failing. I will create an issue for this,  resolve it and push the code

**Comment by najamF18** [comment]
> > > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> > 
> > 
> > I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201
> 
> I have figured out the issue. Actually we are using a rest API for this test. And we should not be using django contrib's LoginRequiredMixin and UserPassesTestMixin. It created a conflict while processing the request and that is why the test is failing. I will create an issue for this, resolve it and push the code

@DeFidelity I have resolved that test case please have a look. And can you figure out why the test cases are failing for postgres? Same test cases are working for sqlite. Please let me know if you have some insights on this.

**Comment by DeFidelity** [comment]
> > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> 
> I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201

So, you think it's all good? besides the test client is already logged in so what could cause the redirection

**Comment by DeFidelity** [comment]
> > > > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> > > 
> > > 
> > > I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201
> > 
> > 
> > I have figured out the issue. Actually we are using a rest API for this test. And we should not be using django contrib's LoginRequiredMixin and UserPassesTestMixin. It created a conflict while processing the request and that is why the test is failing. I will create an issue for this, resolve it and push the code
> 
> @DeFidelity I have resolved that test case please have a look. And can you figure out why the test cases are failing for postgres? Same test cases are working for sqlite. Please let me know if you have some insights on this.

I don't think they're actually failing for Postgres, but I noticed that some other tests are broken now. I'll pull and try to figure things out. Thank you.

**Comment by najamF18** [comment]
> > > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> > 
> > 
> > I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201
> 
> So, you think it's all good? besides the test client is already logged in so what could cause the redirection

Yes. I have resolved this one and the reason for redirection is explained below:
"Actually we are using a rest API for this test. And we should not be using django contrib's LoginRequiredMixin and UserPassesTestMixin. It created a conflict while processing the request and that is why the test is failing. I will create an issue for this, resolve it and push the code"

**Comment by najamF18** [comment]
> > > > > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> > > > 
> > > > 
> > > > I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201
> > > 
> > > 
> > > I have figured out the issue. Actually we are using a rest API for this test. And we should not be using django contrib's LoginRequiredMixin and UserPassesTestMixin. It created a conflict while processing the request and that is why the test is failing. I will create an issue for this, resolve it and push the code
> > 
> > 
> > @DeFidelity I have resolved that test case please have a look. And can you figure out why the test cases are failing for postgres? Same test cases are working for sqlite. Please let me know if you have some insights on this.
> 
> I don't think they're actually failing for Postgres, but I noticed from the logs that some other tests are broken now. I'll pull and try to figure things out. Thank you.

All 6 tests are running successfully in sqllite. But in postgres two tests are failing. Does the database refreshes every time when we create a pull request?

**Comment by DeFidelity** [comment]
> > > > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> > > 
> > > 
> > > I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201
> > 
> > 
> > So, you think it's all good? besides the test client is already logged in so what could cause the redirection
> 
> Yes. I have resolved this one and the reason for redirection is explained below: "Actually we are using a rest API for this test. And we should not be using django contrib's LoginRequiredMixin and UserPassesTestMixin. It created a conflict while processing the request and that is why the test is failing. I will create an issue for this, resolve it and push the code"

Oh! Thanks for the explanations too, do check the log of the failed test here on GitHub to see what caused  the new failure, the test passed before so it might be an error that happened in commit time or maybe a line got deleted from the test file by mistake

**Comment by najamF18** [comment]
> > > > > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> > > > 
> > > > 
> > > > I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201
> > > 
> > > 
> > > So, you think it's all good? besides the test client is already logged in so what could cause the redirection
> > 
> > 
> > Yes. I have resolved this one and the reason for redirection is explained below: "Actually we are using a rest API for this test. And we should not be using django contrib's LoginRequiredMixin and UserPassesTestMixin. It created a conflict while processing the request and that is why the test is failing. I will create an issue for this, resolve it and push the code"
> 
> Oh! Thanks for the explanations too, do check the log of the failed test here on GitHub to see what caused the new failure, the test passed before so it might be an error that happened in commit time or maybe a line got deleted from the test file by mistake

There is no change in tests, I have verified. And all of these tests are running with sqlite. Anyways, I am trying to resolve this.

**Comment by najamF18** [comment]
> > > > > > @najamF18 can you please check if you can find a fix for this failing test? I can't find a reason why the response code keeps returning 200 instead of 201.
> > > > > 
> > > > > 
> > > > > I guess this is because the API returns the webpage to log in via google. That is why it is giving 200 instead of 201
> > > > 
> > > > 
> > > > So, you think it's all good? besides the test client is already logged in so what could cause the redirection
> > > 
> > > 
> > > Yes. I have resolved this one and the reason for redirection is explained below: "Actually we are using a rest API for this test. And we should not be using django contrib's LoginRequiredMixin and UserPassesTestMixin. It created a conflict while processing the request and that is why the test is failing. I will create an issue for this, resolve it and push the code"
> > 
> > 
> > Oh! Thanks for the explanations too, do check the log of the failed test here on GitHub to see what caused the new failure, the test passed before so it might be an error that happened in commit time or maybe a line got deleted from the test file by mistake
> 
> There is no change in tests, I have verified. And all of these tests are running with sqlite. Anyways, I am trying to resolve this.

@DeFidelity I have resolved the issue and pushed the code. Actually while using postgres it was persisting the data created in setUp function.

**Pull request merged**